samples/drivers/gnss/src/main.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2023 Trackunit Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gnss.h>
+
+static const struct device *gnss = DEVICE_DT_GET(DT_ALIAS(gnss));
+
+static const char *gnss_fix_status_to_str(enum gnss_fix_status fix_status)
+{
+	switch (fix_status) {
+	case GNSS_FIX_STATUS_NO_FIX:
+		return "No fix";
+	case GNSS_FIX_STATUS_GNSS_FIX:
+		return "GNSS fix";
+	case GNSS_FIX_STATUS_DGNSS_FIX:
+		return "DGNSS fix";
+	case GNSS_FIX_STATUS_ESTIMATED_FIX:
+		return "Estimated fix";
+	}
+
+	return "Unknown";
+}
+
+static const char *gnss_fix_quality_to_str(enum gnss_fix_quality fix_quality)
+{
+	switch (fix_quality) {
+	case GNSS_FIX_QUALITY_INVALID:
+		return "Invalid";
+	case GNSS_FIX_QUALITY_GNSS_SPS:
+		return "GNSS SPS";
+	case GNSS_FIX_QUALITY_DGNSS:
+		return "DGNSS";
+	case GNSS_FIX_QUALITY_GNSS_PPS:
+		return "GNSS PPS";
+	case GNSS_FIX_QUALITY_RTK:
+		return "RTK";
+	case GNSS_FIX_QUALITY_FLOAT_RTK:
+		return "Float RTK";
+	case GNSS_FIX_QUALITY_ESTIMATED:
+		return "Estimated";
+	}
+
+	return "Unknown";
+}
+
+static void gnss_data_cb(const struct device *dev, const struct gnss_data *data)
+{
+	printk("\r\nData callback\r\n");
+
+	printk("  GNSS Info\r\n");
+	printk("    Satellites count: %u\r\n", data->info.satellites_cnt);
+	printk("    HDOP: %u.%u\r\n", data->info.hdop / 1000, data->info.hdop % 1000);
+	printk("    Fix status: %s\r\n", gnss_fix_status_to_str(data->info.fix_status));
+	printk("    Fix quality: %s\r\n", gnss_fix_quality_to_str(data->info.fix_quality));
+
+	printk("  Navigation data\r\n");
+	printk("    Latitude: %i.%lli degrees\r\n",
+	       (int32_t)(data->nav_data.latitude / 1000000000),
+	       data->nav_data.latitude % 1000000000);
+	printk("    Longitude: %i.%lli degrees\r\n",
+	       (int32_t)(data->nav_data.longitude / 1000000000),
+	       data->nav_data.longitude % 1000000000);
+	printk("    Bearing: %u.%u degrees\r\n", data->nav_data.bearing / 1000,
+	       data->nav_data.bearing % 1000);
+	printk("    Speed: %u.%u m/s\r\n", data->nav_data.speed / 1000,
+	       data->nav_data.speed % 1000);
+	printk("    Altitude: %u.%u m\r\n", data->nav_data.altitude / 1000,
+	       data->nav_data.altitude % 1000);
+
+	printk("  GNSS Time\r\n");
+	printk("    Time: %02u:%02u:%02u.%03u\r\n", data->utc.hour, data->utc.minute,
+	       data->utc.millisecond / 1000, data->utc.millisecond % 1000);
+	printk("    Date: %02u-%02u-%02u\r\n", data->utc.month_day, data->utc.month,
+	       data->utc.century_year);
+}
+
+GNSS_DATA_CALLBACK_DEFINE(DEVICE_DT_GET(DT_ALIAS(gnss)), gnss_data_cb);
```

Comment:
Very cool, have you considered having these pretty print callbacks part of the subsystem code behind a Kconfig? That way you could enable them in any application if you need the data for debugging. Went through this for the input stuff, had a "dump" function part of the sample and found myself copy pasting it all over the place, ended up moving it from the sample into the subsystem and made a shell command for enable/disable dumping if shell is available. Code is https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/input/input_utils.c#L50 if you are interested.

Reply:
I moved the dump logic (what a phrase) to the file gnss_dump.c/h and implemented the dump logic in gnss_publish.c, selectable by the config `CONFIG_GNSS_LOG_DUMP`